man/sortm.man
.TH SORTM %manext1% 2014-03-23 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
sortm \- sort nmh messages
.SH SYNOPSIS
.HP 5
.na
.B sortm
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-all " | " \-noall ]
.RB [ \-datefield
.IR field ]
.RB [ \-textfield
.IR field ]
.RB [ \-notextfield ]
.RB [ \-limit
.IR days ]
.RB [ \-nolimit ]
.RB [ \-check " | " \-nocheck ]
.RB [ \-verbose " | " \-noverbose ]
.ad
.SH DESCRIPTION
.B sortm
sorts the specified messages in the named folder according
to the chronological order of the \*(lqDate:\*(rq field of each message.
.PP
If no messages are specified, the default is all messages in the folder.
However, if no messages are specified and the
.B \-noall
switch is enabled, then
.B sortm
reports that as an error and exits with non-zero status.
.B \-noall
is most useful in the user's profile, to avoid inadvertent sorting of
an entire folder.  The
.B \-all
switch can then be used on the command line to disable it.  Of course,
a message specification of
.I all
can be used in any case.
.PP
The
.B \-verbose
switch directs
.B sortm
to tell the user the general
actions that it is taking to place the folder in sorted order.
.PP
The
.B \-datefield
.I field
switch tells
.B sortm
the name of the field to
use when making the date comparison.  If the user has a special field in
each message, such as \*(lqBB\-Posted:\*(rq or \*(lqDelivery\-Date:\*(rq,
then the
.B \-datefield
switch can be used to direct
.B sortm
which field to examine.
.PP
The
.B \-textfield
.I field
switch causes
.B sortm
to sort messages by the specified text field.  All characters except
letters and numbers are stripped and the resulting strings are sorted
datefield-major, textfield-minor, using a case insensitive
comparison.  If this field is \*(lqsubject\*(rq, any leading "re:" is
stripped off.
.PP
With
.B \-textfield
.IR field ,
if
.B \-limit
.I days
is specified, messages
with the same textfields that are dated within `days' of each other
appear together.  Specifying
.B \-nolimit
makes the limit infinity.
With
.B \-limit
.IR 0 ,
the sort is instead simply textfield-major.
.PP
For example, to order a folder by date-major, grouping messages with
the same subject on the same date together, use:
.PP
.RS 5
sortm -textfield subject +folder
.RE
.PP
.B sortm
always issues a warning for each message that is missing a
\*(lqDate:\*(rq field, has a \*(lqDate:\*(rq field that cannot be
parsed, or has a format error in any header field.  With the
.B \-check
switch,
.B sortm
inhibits all modifications to the folder if there are any such
messages, and exits with non-zero status.  With the default of
.BR \-nocheck ,
.B sortm
sorts messages with a missing or invalid
\*(lqDate:\*(rq field using their file modification times.
.PP
When ordering messages based on their dates, if they have the same
dates, their original message order is preserved.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Current\-Folder:~^To find the default current folder
.fi
.SH "SEE ALSO"
.IR folder (1)
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to the current folder"
.RB ` msgs"' defaults to all without -noall, no default with -noall"
.RB ` \-all '
.RB ` \-datefield "' defaults to date"
.RB ` \-notextfield '
.RB ` \-noverbose '
.RB ` \-nolimit '
.RB ` \-nocheck '
.fi
.SH CONTEXT
If a folder is given, it will become the current folder.  If the current
message is moved,
.B sortm
will preserve its status as current.
.SH HISTORY
Timezones used to be ignored when comparing dates: they aren't any more.
.PP
Messages which were in the folder, but not specified by `msgs', used to
be moved to the end of the folder; now such messages are left untouched.
.PP
.B sortm
sometimes did not preserve the message numbering in a folder
(e.g., messages 1, 3, and 5, might have been renumbered to 1, 2, 3 after
sorting).  This was a bug, and has been fixed.  To compress the message
numbering in a folder, use
.RB \*(lq "folder\ \-pack" \*(rq
as always.
.SH BUGS
When
.B sortm
complains about a message which it can't temporally
order, it complains about the message number
.I prior
to sorting.
It should indicate what the message number will be
.B after
sorting.


NEWS
#
# NEWS - Release notes for nmh 1.6
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly two years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.6 release of nmh:

- All nmh programs will now warn about multiple profile entries for
  the same non-null, non-comment component.  In previous versions, all
  but the first were silently ignored.
- mhmail now supports all post(8) options, or optionally can use
  send(1) with all of its options.  Its formerly undocumented -resent
  switch has been documented.  And it has new -attach, -headerfield,
  and -send/-nosend switches.
- Default to flock() locking on OpenBSD and Darwin.
- Added -directives support to mhbuild(1) to control whether or not
  mhbuild will honor MIME directives by default.  And added support
  for special #on/#off/#pop directives to control the MIME directive
  processing state.
- Added -messageid switch to send(1) and post(8).  This allows
  selection of the style to use for generated Message-ID and
  Content-ID header fields.  The default localname style is
  pid.time@localname, where time is in seconds, and matches previous
  behavior.  The random style replaces the localname with some
  (pseudo)random bytes and uses microsecond-resolution time.
- Added -clobber switch to mhstore(1) to control overwriting of
  existing files.
- Added -outfile switch to mhstore(1).
- Added -noall/-all switches to sortm(1).  sortm -noall requires
  a messages argument.
- $PAGER overrides the compiled-in default pager command.
- Added etc/scan.highlighted format file, as an example of how to
  highlight/colorize the output of scan(1).
- inc(1) now supports a -port switch to specify the port used by the
  POP server.
- pick(1) now decodes MIME-encoded header fields before searching.
- The VISUAL and EDITOR environment variables are now supported as fallbacks
  if the user does not configure an editor entry in their profile.
- The format engine (mh_format(5)) now properly accounts for multibyte
  characters when accounting for column widths.
- burst(1) now can burst MIME-formatted digests (messages that contain
  message/rfc822 parts instead of messages formatted with RFC 934).
- All proc entries (showproc, moreproc, etc) can now accept entries that
  contain spaces and shell metacharacters.  If found, such entries will
  either be space-split or processed by /bin/sh.
- A new program, fmttest(1) is included to help debug format files.  See
  NOTABLE CHANGES below.
- mhshow/mhstore now have support for RFC-2017 (access-type=url) for
  external message bodies.
- Added -retainsequences switch to refile(1).
- A new program, mhfixmsg(1), is included to rewrite MIME messages with
  various transformations.  See NOTABLE CHANGES below.
- Added -[no]rmmproc switches to rmm(1).
- Added support for Content-Disposition header (RFC 2183) to mhstore(1)
  and mhn(1) when used with -auto.
- All nmh commands now support transactional locking for sequence files.
  See NOTABLE CHANGES below.
- There is no longer a per-folder maximum number of sequences.
- For the SMTP MTA TLS can now be negotiated at the beginning of the
  connection with the -initialtls switch.
- Messages can now be selected using a relative offset from some other
  message, or from the start or end of a sequence.  See mh-sequence(5).
- The -changecur and -nochangecur switches have been added to mhlist(1).
- mhbuild(1) can now encode 8-bit message headers using RFC-2047 encoding
  rules.  See NOTABLE CHANGES below.
- The whatnow(1) attach feature will determine the content type of an
  attachment using a program such as file --mime-type, if available at
  configuration time.  If not, it will use mhshow-suffix- entries as
  before.  The -v switch to attach causes it to display the mhbuild
  directive that send(1) will use.
- mhbuild(1) now supports the -auto/-noauto flags (to be used by send(1)
  when invoking mhbuild automatically).
- mhbuild(1) now is automatically run by send, to insure that all outgoing
  messages have proper MIME formatting.  See NOTABLE CHANGES below.
- A new header, "Attach", is supported by mhbuild; it is used to replace
  previous functionality (which by default used a header named
  Nmh-Attachment).  See NOTABLE CHANGES below.
- The default Content-Transfer-Encoding for text parts is now 8bit.
- mhbuild(1) now supports a selectable Content-Transfer-Encoding
- If nmh was configured with iconv(3) support, mhshow will convert, if
  necessary, the charset of text/plain content to match the user's
  locale setting.  See NOTABLE CHANGES below.
- Added support for %{charset} display string escape to mhshow(1).
- The MIME parsing and generating routines now support RFC 2231 extended
  parameter information.
- mh-mime(7) now provides an introduction to nmh's MIME handling.
- mhshow(1) will now by default display all text content under one pager,
  and display markers for non-text and non-inline content.  The content
  markers are changeable via mh-format(5).  See NOTABLE CHANGES below.

-----------------
OBSOLETE FEATURES
-----------------
- Changed exit status of each nmh command's -version and -help
  switches from 1 to 0.
- The following environment variables were deprecated in nmh 1.5
  and removed from this release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature removed)
- Changed repl and dist default to -noatfile.  The default of -atfile

uip/sortm.c
/* sortm.c -- sort messages in a folder by date/time
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/tws.h>
#include <h/utils.h>

#define SORTM_SWITCHES \
    X("datefield field", 0, DATESW) \
    X("textfield field", 0, TEXTSW) \
    X("notextfield", 0, NSUBJSW) \
    X("subject", -3, SUBJSW) /* backward-compatibility */ \
    X("limit days", 0, LIMSW) \
    X("nolimit", 0, NLIMSW) \
    X("verbose", 0, VERBSW) \
    X("noverbose", 0, NVERBSW) \
    X("all", 0, ALLMSGS) \
    X("noall", 0, NALLMSGS) \
    X("check", 0, CHECKSW) \
    X("nocheck", 0, NCHECKSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(SORTM);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(SORTM, switches);
#undef X

struct smsg {
    int s_msg;
    time_t s_clock;
    char *s_subj;
};

static struct smsg *smsgs;
int nmsgs;

char *subjsort = (char *) 0;    /* sort on subject if != 0 */
time_t datelimit = 0;
int submajor = 0;		/* if true, sort on subject-major */
int verbose;
int allmsgs = 1;
int check_failed = 0;

/* This keeps compiler happy on calls to qsort */
typedef int (*qsort_comp) (const void *, const void *);

/*
 * static prototypes
 */
static int read_hdrs (struct msgs *, char *);
static int get_fields (char *, int, struct smsg *);
static int dsort (struct smsg **, struct smsg **);
static int subsort (struct smsg **, struct smsg **);
static int txtsort (struct smsg **, struct smsg **);
static void rename_chain (struct msgs *, struct smsg **, int, int);
static void rename_msgs (struct msgs *, struct smsg **);


int
main (int argc, char **argv)
{
    int	i, msgnum;
    char *cp, *maildir, *datesw = NULL;
    char *folder = NULL, buf[BUFSIZ], **argp;
    char **arguments;
    struct msgs_array msgs = { 0, 0, NULL };
    struct msgs *mp;
    struct smsg **dlist;
    int checksw = 0;

    if (nmh_init(argv[0], 1)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Parse arguments
     */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW:
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW:
		adios (NULL, "-%s unknown", cp);

	    case HELPSW:
		snprintf(buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		print_help (buf, switches, 1);
		done (0);
	    case VERSIONSW:
		print_version(invo_name);
		done (0);

	    case DATESW:
		if (datesw)
		    adios (NULL, "only one date field at a time");
		if (!(datesw = *argp++) || *datesw == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		continue;

	    case TEXTSW:
		if (subjsort)
		    adios (NULL, "only one text field at a time");
		if (!(subjsort = *argp++) || *subjsort == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		continue;

	    case SUBJSW:
		subjsort = "subject";
		continue;
	    case NSUBJSW:
		subjsort = NULL;
		continue;

	    case LIMSW:
		if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		while (*cp == '0')
		    cp++;		/* skip any leading zeros */
		if (!*cp) {		/* hit end of string */
		    submajor++;		/* sort subject-major */
		    continue;
		}
		if (!isdigit((unsigned char) *cp) || !(datelimit = atoi(cp)))
		    adios (NULL, "impossible limit %s", cp);
		datelimit *= 60*60*24;
		continue;
	    case NLIMSW:
		submajor = 0;	/* use date-major, but */
		datelimit = 0;	/* use no limit */
		continue;

	    case VERBSW:
		verbose++;
		continue;
	    case NVERBSW:
		verbose = 0;
		continue;

	    case ALLMSGS:
		allmsgs = 1;
		continue;
	    case NALLMSGS:
		allmsgs = 0;
		continue;

	    case CHECKSW:
		checksw = 1;
		continue;
	    case NCHECKSW:
		checksw = 0;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = pluspath (cp);
	} else
		app_msgarg(&msgs, cp);
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (!msgs.size) {
	if (allmsgs) {
	    app_msgarg(&msgs, "all");
        } else {
	    adios (NULL, "must specify messages to sort with -noall");
        }
    }
    if (!datesw)
	datesw = "date";
    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder, 1)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp, msgs.msgs[msgnum]))
	    done (1);
    seq_setprev (mp);	/* set the previous sequence */

    if ((nmsgs = read_hdrs (mp, datesw)) <= 0)
	adios (NULL, "no messages to sort");

    if (checksw  &&  check_failed) {
	adios (NULL, "errors found, no messages sorted");
    }

    /*
     * sort a list of pointers to our "messages to be sorted".
     */
    dlist = (struct smsg **) mh_xmalloc ((nmsgs+1) * sizeof(*dlist));
    for (i = 0; i < nmsgs; i++)
	dlist[i] = &smsgs[i];
    dlist[nmsgs] = 0;

    if (verbose) {	/* announce what we're doing */
	if (subjsort)
	    if (submajor)
		printf ("sorting by %s\n", subjsort);
	    else
		printf ("sorting by %s-major %s-minor\n", subjsort, datesw);
	else
	    printf ("sorting by datefield %s\n", datesw);
    }

    /* first sort by date, or by subject-major, date-minor */
    qsort ((char *) dlist, nmsgs, sizeof(*dlist),
	    (qsort_comp) (submajor && subjsort ? txtsort : dsort));

    /*
     * if we're sorting on subject, we need another list
     * in subject order, then a merge pass to collate the
     * two sorts.
     */
    if (!submajor && subjsort) {	/* already date sorted */
	struct smsg **slist, **flist;
	struct smsg ***il, **fp, **dp;

	slist = (struct smsg **) mh_xmalloc ((nmsgs+1) * sizeof(*slist));
	memcpy((char *)slist, (char *)dlist, (nmsgs+1)*sizeof(*slist));
	qsort((char *)slist, nmsgs, sizeof(*slist), (qsort_comp) subsort);

	/*
	 * make an inversion list so we can quickly find
	 * the collection of messages with the same subj
	 * given a message number.
	 */
	il = mh_xcalloc(mp->hghsel + 1, sizeof *il);
	for (i = 0; i < nmsgs; i++)
	    il[slist[i]->s_msg] = &slist[i];
	/*
	 * make up the final list, chronological but with
	 * all the same subjects grouped together.
	 */
	flist = (struct smsg **) mh_xmalloc ((nmsgs+1) * sizeof(*flist));
	fp = flist;
	for (dp = dlist; *dp;) {
	    struct smsg **s = il[(*dp++)->s_msg];

	    /* see if we already did this guy */
	    if (! s)
		continue;

	    *fp++ = *s++;
	    /*
	     * take the next message(s) if there is one,
	     * its subject isn't null and its subject
	     * is the same as this one and it's not too
	     * far away in time.
	     */
	    while (*s && (*s)->s_subj[0] &&
		   strcmp((*s)->s_subj, s[-1]->s_subj) == 0 &&
		   (datelimit == 0 ||
		   (*s)->s_clock - s[-1]->s_clock <= datelimit)) {
		il[(*s)->s_msg] = 0;
		*fp++ = *s++;
	    }
	}
	*fp = 0;
	free (il);
	free (slist);
	free (dlist);
	dlist = flist;
    }

    /*
     * At this point, dlist is a sorted array of pointers to smsg structures,
     * each of which contains a message number.
     */

    rename_msgs (mp, dlist);

    context_replace (pfolder, folder);	/* update current folder         */
    seq_save (mp);			/* synchronize message sequences */
    context_save ();			/* save the context file         */
    folder_free (mp);			/* free folder/message structure */
    done (0);
    return 1;
}

static int
read_hdrs (struct msgs *mp, char *datesw)
{
    int msgnum;
    struct smsg *s;

    smsgs = mh_xcalloc(mp->hghsel - mp->lowsel + 2, sizeof *smsgs);
    s = smsgs;
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	if (is_selected(mp, msgnum)) {
	    if (get_fields (datesw, msgnum, s)) {
		s->s_msg = msgnum;
		s++;
	    }
	}
    }
    s->s_msg = 0;
    return(s - smsgs);
}


/*
 * Parse the message and get the data or subject field,
 * if needed.
 */

static int
get_fields (char *datesw, int msg, struct smsg *smsg)
{
    int state;
    int compnum;
    char *msgnam, buf[BUFSIZ], nam[NAMESZ];
    struct tws *tw;
    char *datecomp = NULL, *subjcomp = NULL;
    FILE *in;
    m_getfld_state_t gstate = 0;

    if ((in = fopen (msgnam = m_name (msg), "r")) == NULL) {
	admonish (msgnam, "unable to read message");
	return (0);
    }
    for (compnum = 1;;) {
	int bufsz = sizeof buf;
	switch (state = m_getfld (&gstate, nam, buf, &bufsz, in)) {
	case FLD:
	case FLDPLUS:
	    compnum++;
	    if (!strcasecmp (nam, datesw)) {
		datecomp = add (buf, datecomp);
		while (state == FLDPLUS) {
		    bufsz = sizeof buf;
		    state = m_getfld (&gstate, nam, buf, &bufsz, in);
		    datecomp = add (buf, datecomp);
		}
		if (!subjsort || subjcomp)
		    break;
	    } else if (subjsort && !strcasecmp (nam, subjsort)) {
		subjcomp = add (buf, subjcomp);
		while (state == FLDPLUS) {
		    bufsz = sizeof buf;
		    state = m_getfld (&gstate, nam, buf, &bufsz, in);
		    subjcomp = add (buf, subjcomp);
		}
		if (datecomp)
		    break;
	    } else {
		/* just flush this guy */
		while (state == FLDPLUS) {
		    bufsz = sizeof buf;
		    state = m_getfld (&gstate, nam, buf, &bufsz, in);
		}
	    }
	    continue;

	case BODY:
	case FILEEOF:
	    break;

	case LENERR:
	case FMTERR:
	    if (state == LENERR || state == FMTERR) {
		inform("format error in message %d (header #%d), continuing...",
		      msg, compnum);
		check_failed = 1;
	    }
            mh_xfree(datecomp);
            mh_xfree(subjcomp);
	    fclose (in);
	    return (0);

	default:
	    adios (NULL, "internal error -- you lose");
	}
	break;
    }
    m_getfld_state_destroy (&gstate);

    /*
     * If no date component, then use the modification
     * time of the file as its date
     */
    if (!datecomp || (tw = dparsetime (datecomp)) == NULL) {
	struct stat st;

	inform("can't parse %s field in message %d, "
            "will use file modification time", datesw, msg);
	fstat (fileno (in), &st);
	smsg->s_clock = st.st_mtime;
	check_failed = 1;
    } else {
	smsg->s_clock = dmktime (tw);
    }

    if (subjsort) {
	if (subjcomp) {
	    /*
	     * try to make the subject "canonical": delete
	     * leading "re:", everything but letters & smash
	     * letters to lower case.
	     */
	    char  *cp, *cp2, c;

	    cp = subjcomp;
	    cp2 = subjcomp;
	    if (strcmp (subjsort, "subject") == 0) {
		while ((c = *cp)) {
		    if (! isspace((unsigned char) c)) {
			if(uprf(cp, "re:"))
			    cp += 2;
			else
			    break;
		    }
		    cp++;
		}
	    }

	    while ((c = *cp++)) {
		if (isascii((unsigned char) c) && isalnum((unsigned char) c))
		    *cp2++ = tolower((unsigned char)c);
	    }

	    *cp2 = '\0';
	}
	else
	    subjcomp = "";

	smsg->s_subj = subjcomp;
    }
    fclose (in);
    mh_xfree(datecomp);

    return (1);
}

/*
 * sort on dates.
 */
static int
dsort (struct smsg **a, struct smsg **b)
{
    if ((*a)->s_clock < (*b)->s_clock)
	return (-1);
    if ((*a)->s_clock > (*b)->s_clock)
	return (1);
    if ((*a)->s_msg < (*b)->s_msg)
	return (-1);
    return (1);
}

/*
 * sort on subjects.
 */
static int
subsort (struct smsg **a, struct smsg **b)
{
    int i;

    if ((i = strcmp ((*a)->s_subj, (*b)->s_subj)))
	return (i);

    return (dsort (a, b));
}

static int
txtsort (struct smsg **a, struct smsg **b)
{
    int i;

    if ((i = strcmp ((*a)->s_subj, (*b)->s_subj)))
	return (i);
    if ((*a)->s_msg < (*b)->s_msg)
	return (-1);
    return (1);
}

static void
rename_chain (struct msgs *mp, struct smsg **mlist, int msg, int endmsg)
{
    int nxt, old, new;
    char *newname, oldname[BUFSIZ];
    char newbuf[PATH_MAX + 1];

    for (;;) {
	nxt = mlist[msg] - smsgs;	/* mlist[msg] is a ptr into smsgs */
	mlist[msg] = NULL;
	old = smsgs[nxt].s_msg;
	new = smsgs[msg].s_msg;
	strncpy (oldname, m_name (old), sizeof(oldname));
	newname = m_name (new);
	if (verbose)
	    printf ("message %d becomes message %d\n", old, new);

	(void)snprintf(oldname, sizeof (oldname), "%s/%d", mp->foldpath, old);
	(void)snprintf(newbuf, sizeof (newbuf), "%s/%d", mp->foldpath, new);
	ext_hook("ref-hook", oldname, newbuf);

	if (rename (oldname, newname) == NOTOK)
	    adios (newname, "unable to rename %s to", oldname);

	copy_msg_flags (mp, new, old);
	if (mp->curmsg == old)
	    seq_setcur (mp, new);

	if (nxt == endmsg)
	    break;

	msg = nxt;
    }
/*	if (nxt != endmsg); */
/*	rename_chain (mp, mlist, nxt, endmsg); */
}

static void
rename_msgs (struct msgs *mp, struct smsg **mlist)
{
    int i, j, old, new;
    bvector_t tmpset = bvector_create (0);
    char f1[BUFSIZ], tmpfil[BUFSIZ];
    char newbuf[PATH_MAX + 1];
    struct smsg *sp;

    strncpy (tmpfil, m_name (mp->hghmsg + 1), sizeof(tmpfil));

    for (i = 0; i < nmsgs; i++) {
	if (! (sp = mlist[i]))
	    continue;   /* did this one */

	j = sp - smsgs;
	if (j == i)
	    continue;   /* this one doesn't move */

	/*
	 * the guy that was msg j is about to become msg i.
	 * rename 'j' to make a hole, then recursively rename
	 * guys to fill up the hole.
	 */
	old = smsgs[j].s_msg;
	new = smsgs[i].s_msg;
	strncpy (f1, m_name (old), sizeof(f1));

	if (verbose)
	    printf ("renaming message chain from %d to %d\n", old, new);

	/*
	 *	Run the external hook to refile the old message as the
	 *	temporary message number that is off of the end of the
	 *	messages in the folder.
	 */

	(void)snprintf(f1, sizeof (f1), "%s/%d", mp->foldpath, old);
	(void)snprintf(newbuf, sizeof (newbuf), "%s/%d", mp->foldpath, mp->hghmsg + 1);
	ext_hook("ref-hook", f1, newbuf);

	if (rename (f1, tmpfil) == NOTOK)
	    adios (tmpfil, "unable to rename %s to ", f1);

	get_msg_flags (mp, tmpset, old);

	rename_chain (mp, mlist, j, i);

	/*
	 *	Run the external hook to refile the temorary message number
	 *	to the real place.
	 */

	(void)snprintf(f1, sizeof (f1), "%s/%d", mp->foldpath, new);
	ext_hook("ref-hook", newbuf, f1);

	if (rename (tmpfil, m_name(new)) == NOTOK)
	    adios (m_name(new), "unable to rename %s to", tmpfil);

	set_msg_flags (mp, tmpset, new);
	mp->msgflags |= SEQMOD;
    }

    bvector_free (tmpset);
}
